Dirt-road backup & restore

ch1bo opened the pull request:
Gave it a shot of saving / loading both `HeadState` and `ChainState`. This is not using ADR18.

#187 

To check before merging:
* [x] CHANGELOG is up to date
* [x] Documentation

pgrange committed:
WIP intitializing high level backup restore test

v0d1ch committed:
Make the test fail without persistence in place

ch1bo committed:
Draft where to write the HeadState to disk

ch1bo committed:
Introduce a Persistence handle and use it to save state

ch1bo committed:
Make loading optional

ch1bo committed:
Check for file existence and trace what was loaded

ch1bo committed:
Use Persistence handle in chain layer

ch1bo committed:
Derive To/FromJSON instances for chain state

ch1bo committed:
Add orphan json instances for cardano api PlutusScript

ch1bo committed:
Add To/FromJSON for onchain Party type using serialise (cbor)

ch1bo committed:
Add orphan To/FromJSON instance for cardano api script data

ch1bo committed:
Use orphan instances to get To/FromJSON of ChainState

ch1bo committed:
Add orphan FromJSON Plutus.POSIXTime

ch1bo committed:
Add orphan To/FromJSON instances for VerificationKey and NetworkId

ch1bo committed:
Derive more instances

ch1bo committed:
Introduce UTxOHash type in State/Tx modules

Required to make To/FromJSON instances for this bytestring

ch1bo committed:
Add orphan To/FromJSON instances for ChainPoint

ch1bo committed:
Make tests compile

ch1bo committed:
Use base16 to fix To/FromJSON instances

Also don't ignore faulty files as this is annoying to discover them in
the repository.

ch1bo committed:
Fix end-to-end spec assertion -> green :green_heart:

ch1bo committed:
Make persistence directory configurable via command line

ch1bo committed:
Add log of loaded chain state

ch1bo committed:
Write using unliftio's more robust function

Also need to check whether file is empty for some reason.

ch1bo committed:
Ignore rollback past IdleState as this can actually happen now

ch1bo committed:
Remove focus on e2e tests

ch1bo committed:
Fix a warning in OptionsSpec

ch1bo committed:
Remove a redundant import

ch1bo committed:
Add --persistence-dir to toArgs

ch1bo committed:
Update logs schema

ch1bo committed:
Use state-specific persistence dirs in e2e tests

ffakenz reviewed (approved):
LGTM

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -178,6 +179,7 @@ library
     , typed-protocols
     , typed-protocols-cborg
     , typed-protocols-examples
+    , unliftio
```
Used `writeBinaryFileDurableAtomic` from there to ensure files are written (it writes to a temp file and moves it after write, which is atomic).

ch1bo commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -78,8 +79,9 @@ spec = around showLogsOnFailure $ do
-          withDirectChain (contramap (FromDirectChain "alice") tracer) networkId iocp nodeSocket aliceKeys alice cardanoKeys Nothing hydraScriptsTxId (putMVar alicesCallback) $ \Chain{postTx} -> do
-            withDirectChain nullTracer networkId iocp nodeSocket bobKeys bob cardanoKeys Nothing hydraScriptsTxId (putMVar bobsCallback) $ \_ -> do
+          persistence <- createPersistence Proxy $ tmp <> "/chainstate"
+          withDirectChain (contramap (FromDirectChain "alice") tracer) networkId iocp nodeSocket aliceKeys alice cardanoKeys Nothing hydraScriptsTxId persistence (putMVar alicesCallback) $ \Chain{postTx} -> do
```
Yes

v0d1ch commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DuplicateRecordFields #-}
 {-# LANGUAGE NumericUnderscores #-}
+{-# OPTIONS_GHC -Wno-unused-do-bind #-}
```
I guess this is here by accident

v0d1ch reviewed (approved):
Nice work!

ch1bo committed:
Update README and CHANGELOG with info about new persistence

ch1bo committed:
Include a verbatim output of the hydra-node --help in quickstart

This avoids maintaining two sources of documenting the individual
command line options.

ch1bo committed:
Remove redundant warning pragma

ch1bo commented:
Re-requesting reviews as I have updated the documentation now with a verbatim copy of `hydra-node --help` to avoid duplication. @v0d1ch @ffakenz @abailly-iohk

v0d1ch commented on docs/docs/getting-started/quickstart.md:
```diff
@@ -20,28 +20,106 @@ This image contains both `cardano-node` and `cardano-cli`. The latter is handy t
-`--monitoring-port`             | The port this node listens on for monitoring and metrics via Prometheus.<br />If left empty, monitoring server is not started.
-`--network-id`                  | The id of a test network to conect, with the given network magic.
-`--start-chain-from`            | The id of the block we want to start observing the ledger from.<br />Composed by the slot number, a separator (".") and the hash of the block header. For example: 52970883.d36a9936ae7a07f5f4bdc9ad0b23761cb7b14f35007e54947e27a1510f897f04
+The entire configuration of the `hydra-node` is provided through command-line options. Options are used to configured various elements of the network, API, chain connection and used ledger. You can use the `--help` option to get a description of all options:
```
`configured -> configure`

v0d1ch commented on hydra-cardano-api/src/Hydra/Cardano/Api/NetworkId.hs:
```diff
@@ -0,0 +1,26 @@
+    case tag :: Text of
+      "Mainnet" -> pure Mainnet
+      "Testnet" -> Testnet . NetworkMagic <$> o .: "magic"
+      _ -> fail "Expected tag to be Mainnet | Tesnet"
```
`Tesnet -> Testnet`

v0d1ch commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -240,7 +263,43 @@ putState :: HydraHead tx m -> HeadState tx -> STM m ()
+          liftIO (doesFileExist fp) >>= \case
+            False -> pure Nothing
+            True -> do
+              bs <- toStrict <$> readFileLBS fp
```
Can we just use `readFileBS` to avoid a call to `toStrict` here?

v0d1ch reviewed (approved):
Found couple of typos, lgtm :tada:

ch1bo commented:
@abailly-iohk not sure why, but I can't reply on the comments. Re the lwn article: 

The sequence of steps in the "when to fsync" section is actually what the unliftio function does. Is adding a dependency for it worth it? I'd say yes (over vendoring).. until we do something else than write json to disk. An objective argument against could be performance, but I wouldn't know as I did not run any benchmarks.

ghost commented:
I would personally have just done `BS.writeFile ...` because of the "dirt road" aspect of this feature and develop a better storage method on top of sqlite. The linked article is not actually the one I had in mind, which is a deep technical explanation of the things sqlite needs to do to ensure data is actually written on disk. Or perhaps was it postgresql? Anyway, the conclusion I drew from this is that it's much much safer to rely on this battle-tested and simple to use DB rather than trying to use Haskell's IO when one wants to provide strong durability guarantees.
That said I have been using basic file IO in production at Capital Match and never had an issue (except when we were not writing to disk but that's another story). Of course, load was rather small, with few spikes (can't remember the numbers but perhaps we had some 50tps in peak?).

ch1bo committed:
Revert change of default --host to 127.0.0.1

Let's make this breaking change another time.

ch1bo committed:
Fix a typo

ch1bo committed:
Fix quickstart instructions and also update translated content..somewhat

ch1bo committed:
Use readFileBS instead of toStrict <$> readFileLBS

ch1bo committed:
Do not include loaded state in logs

We decided that this is not as helpful and requires lots of schema
definitions + bloats the logs.

ch1bo committed:
Also do not log when 'SavingState'

This is not even having any payload and we can infer that this was tried
by the event handling / resulting in a new state.

ch1bo committed:
Add CreatedState log to indicate when nothing was loaded from persistence
